thrust/thrust/system/cuda/detail/core/agent_launcher.h

```diff
@@ -70,609 +70,30 @@ namespace core
-{
-  extern __shared__ char shmem[];
-  vshmem = vshmem == NULL ? shmem : vshmem + blockIdx.x * temp_storage_size<typename Agent::ptx_plan>::value;
-  Agent::entry(x0, x1, x2, vshmem);
-}
-template <class Agent, class _0, class _1, class _2, class _3>
-THRUST_DETAIL_KERNEL_ATTRIBUTES void __launch_bounds__(Agent::ptx_plan::BLOCK_THREADS)
-  _kernel_agent_vshmem(char* vshmem, _0 x0, _1 x1, _2 x2, _3 x3)
-{
-  extern __shared__ char shmem[];
-  vshmem = vshmem == NULL ? shmem : vshmem + blockIdx.x * temp_storage_size<typename Agent::ptx_plan>::value;
-  Agent::entry(x0, x1, x2, x3, vshmem);
-}
-template <class Agent, class _0, class _1, class _2, class _3, class _4>
-THRUST_DETAIL_KERNEL_ATTRIBUTES void __launch_bounds__(Agent::ptx_plan::BLOCK_THREADS)
-  _kernel_agent_vshmem(char* vshmem, _0 x0, _1 x1, _2 x2, _3 x3, _4 x4)
-{
-  extern __shared__ char shmem[];
-  vshmem = vshmem == NULL ? shmem : vshmem + blockIdx.x * temp_storage_size<typename Agent::ptx_plan>::value;
-  Agent::entry(x0, x1, x2, x3, x4, vshmem);
-}
-template <class Agent, class _0, class _1, class _2, class _3, class _4, class _5>
-THRUST_DETAIL_KERNEL_ATTRIBUTES void __launch_bounds__(Agent::ptx_plan::BLOCK_THREADS)
-  _kernel_agent_vshmem(char* vshmem, _0 x0, _1 x1, _2 x2, _3 x3, _4 x4, _5 x5)
-{
-  extern __shared__ char shmem[];
-  vshmem = vshmem == NULL ? shmem : vshmem + blockIdx.x * temp_storage_size<typename Agent::ptx_plan>::value;
-  Agent::entry(x0, x1, x2, x3, x4, x5, vshmem);
-}
-template <class Agent, class _0, class _1, class _2, class _3, class _4, class _5, class _6>
-THRUST_DETAIL_KERNEL_ATTRIBUTES void __launch_bounds__(Agent::ptx_plan::BLOCK_THREADS)
-  _kernel_agent_vshmem(char* vshmem, _0 x0, _1 x1, _2 x2, _3 x3, _4 x4, _5 x5, _6 x6)
-{
-  extern __shared__ char shmem[];
-  vshmem = vshmem == NULL ? shmem : vshmem + blockIdx.x * temp_storage_size<typename Agent::ptx_plan>::value;
-  Agent::entry(x0, x1, x2, x3, x4, x5, x6, vshmem);
-}
-template <class Agent, class _0, class _1, class _2, class _3, class _4, class _5, class _6, class _7>
-THRUST_DETAIL_KERNEL_ATTRIBUTES void __launch_bounds__(Agent::ptx_plan::BLOCK_THREADS)
-  _kernel_agent_vshmem(char* vshmem, _0 x0, _1 x1, _2 x2, _3 x3, _4 x4, _5 x5, _6 x6, _7 x7)
-{
-  extern __shared__ char shmem[];
-  vshmem = vshmem == NULL ? shmem : vshmem + blockIdx.x * temp_storage_size<typename Agent::ptx_plan>::value;
-  Agent::entry(x0, x1, x2, x3, x4, x5, x6, x7, vshmem);
-}
-template <class Agent, class _0, class _1, class _2, class _3, class _4, class _5, class _6, class _7, class _8>
+template <class Agent, class... Args>
 THRUST_DETAIL_KERNEL_ATTRIBUTES void __launch_bounds__(Agent::ptx_plan::BLOCK_THREADS)
-  _kernel_agent_vshmem(char* vshmem, _0 x0, _1 x1, _2 x2, _3 x3, _4 x4, _5 x5, _6 x6, _7 x7, _8 x8)
+  _kernel_agent_vshmem(char* vshmem, Args... args)
 {
   extern __shared__ char shmem[];
   vshmem = vshmem == NULL ? shmem : vshmem + blockIdx.x * temp_storage_size<typename Agent::ptx_plan>::value;
```

Comment:
While we are at it, can we drop the `NULL` too?

Reply:
There you go: #1805